### PR TITLE
[OHFJIRA-106] : Circuit breaker

### PR DESCRIPTION
**Pull request by kkovarik**
## Circuit breaker

Documentation : https://openhubframework.atlassian.net/wiki/spaces/OHF/pages/1177288705/circuit
Please let me know what parts of it should be improved in any way, thanks!

Part of this pull request is camel component itself & two implementations of CircuitBreaker interface. Both are mentioned in documentation, one is in-memory, second is hazelcast based. There are three commits, as it may be quite a lot of code for one PR.

### Open questions
* name of the component - should it be `circuit` or `circuit-breaker`? currently it is circuit, to avoid boiler plate, as while used it can easily become quite unreadable: `circuit:my-system:http4://www.my-system.com/resource?method=POST&timeout=40000...`
* shouldn´t the configuration of component be more "camel-like", meaning it would parse uri and not to use the exchange property?

### Possible improvements
* currently only one implementation of CircuitBreaker interface is active for all circuits, technically it could be modified, that each circuit could have its own

### CODE REVIEW

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.openhubframework.openhub.spi.circuitbreaker;
+
+/**
+ * Exception to be thrown, when circuit is down (shortcuted).
+ *
+ * @author Karel Kovarik
+ * @since 2.2
+ */
+public class CircuitDownException extends RuntimeException {
```
Comment (pjuza) [inline core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java]: Muze to dedit od IntegrationException?

File: core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/CircuitState.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * CircuitState domain object, does store circuit state.
+ *
+ * @author Karel Kovarik
+ * @since 2.2
+ */
+public class CircuitState implements Serializable {
+    static final long serialVersionUID = 1L;
+
+    private long lastShortcutTimestamp;
+    private List<Long> successCallList = new CopyOnWriteArrayList<>();
```
Comment (pjuza) [inline core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/CircuitState.java]: Neni to zbytecny komfort pouzivat CopyOnWriteArrayList? Chapu, ze ti to vyresi problem s pristupem pres get() ke kolekci, ale trpi tim performance a mozna zbytecne. Davam k uvaze ...

File: components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    @Autowired
+    private CamelContext camelContext;
+
+
+    @Override
+    protected Endpoint createEndpoint(String uri, String remaining, Map<String, Object> parameters) throws Exception {
+        final CircuitEndpoint endpoint = new CircuitEndpoint(uri, this);
+
+        final String endpointURI = ObjectHelper.after(uri, ":");
```
Comment (pjuza) [inline components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java]: IDE me ukazuje, ze ObjectHelper je deprecated, ze mas pouzit StringHelper

File: components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package org.openhubframework.openhub.component.circuitbreaker;
```
Comment (pjuza) [inline components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java:1]: pridal bych sem i do jinych novych baliku packe-info.java

File: components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitProducer.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    public CircuitProducer(CircuitEndpoint endpoint) {
+        super(endpoint);
+    }
+
+    @Override
+    public void process(Exchange exchange) throws Exception {
+        // verify configuration is present, otherwise fail as it is misconfigured
+        final CircuitConfiguration circuitConfiguration = exchange.getProperty(
+                CircuitBreaker.CONFIGURATION_PROPERTY, CircuitConfiguration.class);
+        Assert.notNull(circuitConfiguration, "the circuitConfiguration was not set.");
```
Comment (pjuza) [inline components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitProducer.java]: Asi bych jeste pridal info, ze neni nastaveno v Exchange + nazev te property, protoze preci jen to je trochu nestandardni, takze tato chyba bude vylitavat casto :)

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+
+/**
+ * Contract for Circuit Breaker.
+ *
+ * Recommended to be used with org.openhubframework.openhub.component.circuitbreaker.CircuitComponent,
+ * see its javadoc.
+ *
+ * Usage:
+ * .doTry()
```
Comment (pjuza) [inline core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java:15]: chtelo by to nejake formatovace do javaDoc, jinak je pak vse na jednom radku.

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * Get current circuit configuration.
+     *
+     * @param exchange the camel exchange.
+     * @return the circuit configuration.
+     */
+    default CircuitConfiguration getCircuitConfiguration(Exchange exchange) {
+        final CircuitConfiguration circuitConfiguration =
+                exchange.getProperty(CONFIGURATION_PROPERTY, CircuitConfiguration.class);
+        Assert.notNull(circuitConfiguration, "the circuitConfiguration was not found");
```
Comment (pjuza) [inline core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java]: tady bych tu chybovou hlasku take rozsiril ...

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * If set to false, circuit breaker is disabled, meaning all calls will
+     * be invoked "live".
+     */
+    private boolean enabled = true;
+    /**
+     * Unique name of circuit.
+     */
+    private String circuitName;
```
Comment (pjuza) [inline core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitConfiguration.java:24]: zadny konstruktor? mozna jemn nastavit nejaky default circuitName ...

File: core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/AbstractCircuitBreaker.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
+package org.openhubframework.openhub.core.circuitbreaker;
```
Comment (pjuza) [inline core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/AbstractCircuitBreaker.java:1]: pridal bych package-info.java

**Review by pjuza — CHANGES_REQUESTED** [review]

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java (hunk 1 above)
Comment (kkovarik) [reply core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java]: Tak asi by mohlo. IntegrationException by měly být všechny vyjímky z "vnitřku" OpenHubu? :)

File: core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/CircuitState.java (hunk 2 above)
Comment (kkovarik) [reply core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/CircuitState.java]: Dobrá připomínka, vzniklo to při implementaci InMemory varianty. Možná opravdu zbytečné, upravím a doplním komentářem

File: components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java (hunk 3 above)
Comment (kkovarik) [reply components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java]: upraveno, díky

File: components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java (hunk 4 above)
Comment (kkovarik) [reply components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitComponent.java:1]: Díky za upozornění, přidal jsem kde jsem našel

File: components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitProducer.java (hunk 5 above)
Comment (kkovarik) [reply components/src/main/java/org/openhubframework/openhub/component/circuitbreaker/CircuitProducer.java]: Určitě, dobrý tip 👍 
Doplněno.

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java (hunk 6 above)
Comment (kkovarik) [reply core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java:15]: upraveno, díky za upozornění

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java (hunk 7 above)
Comment (kkovarik) [reply core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitBreaker.java]: done

File: core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/AbstractCircuitBreaker.java (hunk 9 above)
Comment (kkovarik) [reply core/src/main/java/org/openhubframework/openhub/core/circuitbreaker/AbstractCircuitBreaker.java:1]: přidáno

File: core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java (hunk 1 above)
Comment (pjuza) [reply core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java]: Ano, myslim, ze takova byla/je puvodni predstava. https://openhubframework.atlassian.net/wiki/spaces/OHF/pages/33595/Error+handling

Comment (kkovarik) [reply core-spi/src/main/java/org/openhubframework/openhub/spi/circuitbreaker/CircuitDownException.java]: Supr, díky, upraveno :)

**Review by pjuza — APPROVED** [review]
Super, thank you for changes.

Please look at Sonar issues (https://sonarcloud.io/dashboard?id=OpenWiseSolutions_openhub-framework&pullRequest=101) if something is useful for you before final finishing.

**Pull request merged**